temporian/implementation/numpy/operators/window/base.py

```diff
@@ -0,0 +1,121 @@
+    window operators."""
+
+    def __init__(self, op: BaseWindowOperator) -> None:
+        super().__init__()
+        assert isinstance(op, BaseWindowOperator)
+        self._op = op
+
+    def __call__(
+        self,
+        event: NumpyEvent,
+        sampling: Optional[NumpyEvent] = None,
+    ) -> Dict[str, NumpyEvent]:
+        if sampling is None:
+            sampling = event
+
+        dst_event = NumpyEvent(data={}, sampling=sampling.sampling)
+
+        # Naming convention:
+        #   mts => multi time series
+        #   ts => time series
+        #   dst => destination
+        #   src => source
+
+        # For each index
+        for index, src_mts in event.data.items():
+            dst_mts = []
+            dst_event.data[index] = dst_mts
+            src_timestamps = event.sampling.data[index]
+            sampling_timestamps = sampling.sampling.data[index]
+
+            mask = self._build_accumulator_mask(
+                src_timestamps, sampling_timestamps, self._op.window_length()
+            )
+
+            # For each feature
+            for src_ts in src_mts:
+                dst_feature_name = (
+                    f"{self._op.output_feature_prefix}_{src_ts.name}"
+                )
+                dst_ts_data = self._apply_accumulator_mask(src_ts.data, mask)
+                dst_mts.append(NumpyFeature(dst_feature_name, dst_ts_data))
+
+        return {"event": dst_event}
+
+    def _build_accumulator_mask(
+        self,
+        data_timestamps: np.array,
+        sampling_timestamps: np.array,
+        win: Duration,
+    ) -> np.array:
+        """Creates a 2d boolean matrix containing the summing instructions.
+
+        The returned matrix "m[i,j]" is defined as:
+            m[i,j] is true iif. input value "j" is averaged in the output value "i".
+        """
+
+        # This implementation is simple but expensive. It will create multiple
+        # O(n^2) arrays, where n is the number of time samples.
+
+        right_side = (
+            sampling_timestamps[:, np.newaxis] >= data_timestamps[np.newaxis, :]
+        )
+
+        # TODO: Make left side inclusivity/exclusivity a parameter.
+        left_side = sampling_timestamps[:, np.newaxis] <= (
+            data_timestamps[np.newaxis, :] + win
+        )
+
+        return right_side & left_side
+
+    def _apply_accumulator_mask(
+        self, src: np.array, mask: np.array
+    ) -> np.array:
+        """Sums elements according to an accumulator mask."""
+
+        # Broadcast of feature values to requested timestamps.
+        cross_product = src * mask
+
+        # Replace masked values with NaN
+        cross_product[
+            mask == False  # pylint: disable=singleton-comparison
+        ] = np.nan
+
+        return self._calculate_window_operation(cross_product)
+
+    @abstractmethod
+    def _calculate_window_operation(self, values: np.array) -> np.array:
+        """Calculates the window operation on the given values.
```

Comment:
Would be handy to know from the docstring what shape `values` is expected to have and what shape is expected as output (since both are arrays and it isn't obvious from the code nor the docstrings what they are), or at least that the output array should be same as input but with one collapsed dimension

Reply:
also - I feel the name `calculate_window_operation` could be improved here, since the window logic is done by `_build_accumulator_mask` and `_apply_accumulator_mask`, this function is just applying the operation (operation = {average, std}). maybe `_apply_operation`? and making it clear in the docstring that this shouldn't be doing any window logic (would change line 120 for example)